riscv-configuration-structure-draft.adoc

```diff
@@ -268,9 +268,37 @@ big immediate benefit of a format that supports comments.
 [[sec:AccessMethod]]
 == Access Method
 
-The binary configuration structure is memory-mapped in system memory.
-There is a CSR which contains the physical address where the structure
-starts.
+The binary configuration structure is memory-mapped in system memory.  There is
+a CSR which contains the physical address where the structure starts.
+
+TODO: Get a CSR number assigned.
+
+When software running on a hart wants to read the configuration structure, it
+reads this CSR, and then decodes the binary structure at that physical address.
+The structure will specify which parts apply to which harts (identified by hart
+ID), and the software can ignore any information that does not apply to the hart
+it's running on.
+
+This provides hardware implementers several options for embedding the binary
+structure:
+1. The structure can describe all harts, and be accessible over the memory bus.
+All harts have the same memory map and the same CSR contents.
+2. There might be several structures in the system, and different harts are
+pointed to different structures by having different memory maps.
+3. There might be several structures in the system, and different harts are
+pointed to different structures by having different pointers in their CSR.
+4. A combination of 2 and 3 above could be used.
+
+Hardware implementers have a lot of flexibility to handle everything from simple
+fixed systems to complex socketed systems. In each case it's straightforward to
+ensure that each hart can read a configuration structure that describes its own
+capabilities.
+
+How to discover information about each hart in a socketed system is an open
+question. Presumably each chip contains a structure it describes itself. It
+would be nice to have some definition of a motherboard configuration structure,
+which could contain pointers to the configuration structures in each chip.  That
+feels outside the scope of what the RISC-V Foundation defines, however.
```

Comment:
Agree with this, the platform may provide the hardware strapping to each socket, so the processor on each socket can determine the memory map address of CS without conflicts.

Reply:
Thinking of your comment above, we could add an optional list of pointers to other structures to the schema. That way systems with a universal memory map but different CSR pointers per chip can get info about all the components in the system.

Reply:
Not quite understand this method. Can this method support multiple sockets but only one processor is installed? The second socket may be installed with an identical or different SKU processor later in the next power cycle?

Reply:
Assume a motherboard with 2 sockets. Let's say there's this memory map (implemented on the motherboard):
0x8000: structure containing pointers to 0x10000 and 0x11000
0x10000: points to the configuration structure embedded in the chip inserted into socket 0.
0x11000: points to the configuration structure embedded in the chip inserted into socket 1.

Each chip to be inserted into such a system has mconfigptr set to 0x8000, and contains a configuration structure that describes its own harts.

So... firmware runs. It reads mconfigptr, looks at 0x8000, and then looks at 0x10000 and 0x11000, and now has a complete picture of the system. (I'm not sure how to deal with the fact that each chip should have unique hart IDs. You can have chips know which socket they are in and adjust hart ID, but you can't really have them adjust the configuration structure because the structure is too complex.)

Reply:
Ok, instead of a boolean, a pointer to a list works as well.
But we don't need to have a list entry that points to another socket. The table of lists is only for the local processor package. The processor itself should be socket agnostic. A table of lists that point to CSs of each local hart makes sense to me.

Below is the flow of FW reads the CS, when each hart in each physical processor boots up
FW reads the value from mconfigptr 
   - If it returns 0, then means no there is no CS implemented in this hart.
   - If non zero then FW read the table of lists. It could be a one-entry list or multiple list entries which depends on the implementation. (the size of the table or the number of lists is TBD)
   - FW read the CS from each list.
 FW can then aggregates all CS it reads from harts.

 Above can be applied on below CS implementations,
- One hart describes the entire CS of all harts in the processor package.
- Each hart describes its own CS.
- Or mixup

The address of CS could be,
 - the local (private) address to the hart
 - the global address to each harts.

Reply:
I'll make a separate PR for adding pointers to the structure.

Reply:
Actually, I already had added it to this PR. See structurePointers in Configuration.